### DISCUSSION TITLE
Guide for writing custom dev service

### DISCUSSION THREAD

**Discussion by joggeli34** [quarkusio/quarkus#26791] [Q&A]
Hi,

We like to write some custom dev services. For example we like to have following dev services:
* InfluxDB
* MQTT (v5)
* An extension for the keycloak dev-services which runs a docker container after start to initialize a custom realm

It would be nice to have a guide for writing custom dev services. I didn't find one, so my question is if I just missed it?
Or is it just not recommended to write custom dev services?

Regards

**Comment by sberyozkin** [comment]
@joggeli34 Hi, my answer is off-topic, but re

> An extension for the keycloak dev-services which runs a docker container after start to initialize a custom realm

`Dev Services for Keycloak` posts custom realm files to initialize the realm and it has been tested. So it should work, but if it does not then please start a new discussion or open an issue. Perhaps we can somehow support customizing `KeycloakDevServicesProcessor` without the users having to create a new extension or add a build item notifying the interested extensions that `KeycloakDeServicesProcessor` has completed its work.

**Reply by joggeli34** [reply]
We are currently not using realm files, because we use the keycloak-admin-client to add the realm running in an separate container. This allows us to not only create new realms but only update and migrate them (somehow similar to a liquibase of flyway script for the db). And it would be nice if we could use this container instead of maintaining an additional realm.json which then has to be synchornized somehow.
So for this use-case it would be great to have some callback or something from the `KeycloakDevServicesProcessor` which indicates that he has finished, as you suggested.

**Comment by holly-cummins** [comment]
Resolved by https://github.com/quarkusio/quarkus/pull/42535. The guide is live at https://quarkus.io/version/main/guides/extension-writing-dev-service.
